compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -775,45 +775,77 @@ static SmallVector<int64_t> getDefaultMatmulCacheSizes(linalg::LinalgOp op,
-  // The tiling for parallel dims and reduction dims should be separated.
-  SmallVector<int64_t> parallelTileSizes(vecTileSizes.begin(),
-                                         vecTileSizes.end());
-  parallelTileSizes.back() = 0;
-
-  // Clamp inner tiling sizes to avoid masking. The vector masking takes the
-  // last level of tiling to create masks. It would lead to incorrect masking if
-  // the inner tiling sizes are not clamped. Because padding won't be applied
-  // along those dimensions.
-  for (const auto &[index, size] : llvm::enumerate(distTileSizes)) {
-    if (!size)
-      continue;
-    parallelTileSizes[index] = std::min(parallelTileSizes[index], size);
-  }
 
-  // TODO(hanchung): Make logic more heuristic. Peeling hurts performance a lot
-  // if the dim size is small (e.g., K=24).
+  // 1. Compute tile sizes for all tiling levels.
+  // The tiling for parallel dims (M and N) and reduction dim (K) should be
+  // separated, so we move K dim from parallel tile sizes to reduction tile
+  // sizes.
   int64_t numTilingDims = vecTileSizes.size();
-  SmallVector<int64_t> reductionTileSizes(numTilingDims - 1, 0);
-  auto lhsShapedType = llvm::cast<ShapedType>(op.lhs().getType());
-  int64_t K = lhsShapedType.getShape().back();
-  reductionTileSizes.push_back(
-      getMaxVectorTileSize(K, vecTileSizes.back(), vectorSize));
-
   SmallVector<int64_t> cacheParallelTileSizes(cacheTileSizes.begin(),
                                               cacheTileSizes.end());
   SmallVector<int64_t> cacheReductionTileSizes(numTilingDims, 0);
   std::swap(cacheParallelTileSizes.back(), cacheReductionTileSizes.back());
 
+  SmallVector<int64_t> vectorParallelTileSizes(vecTileSizes.begin(),
+                                               vecTileSizes.end());
+  SmallVector<int64_t> vectorReductionTileSizes(numTilingDims, 0);
+  std::swap(vectorParallelTileSizes.back(), vectorReductionTileSizes.back());
+
   TileSizesListType tileSizes = {
       SmallVector<int64_t>(distTileSizes), cacheParallelTileSizes,
-      cacheReductionTileSizes, parallelTileSizes, reductionTileSizes};
-
+      cacheReductionTileSizes, vectorParallelTileSizes,
+      vectorReductionTileSizes};
   // No need for tiling inner parallel dims.
   tileSizes.emplace_back(numTilingDims, 0);
 
+  // The backend struggles to legalize non-power-of-two scalable vectors,
+  // hence the extra rounding up.
+  const SmallVectorImpl<bool> &vecScalableDims = inputScalableTileFlags.back();
+  for (const auto &[index, size] : llvm::enumerate(vectorParallelTileSizes)) {
+    if (!size)
+      continue;
+    vectorParallelTileSizes[index] =
+        roundUpToPow2(size,
+                      /*predicate=*/vecScalableDims[index]);
+  }
+
+  // 2. Set scalable flags for all the tiling levels.
+  SmallVector<bool> parallelScalableFlags;
+  for (auto [index, tileSize] : llvm::enumerate(vecTileSizes))
+    parallelScalableFlags.push_back(vecScalableDims[index]);
```

Comment:
I think this can just be

```
SmallVector<bool> parallelScalableFlags = vecScalableDims;
```

Did I miss something?

Reply:
That particular constructor is `delete`d, but you are right that this can and should be simplified :)